scripts/magpietts/update_magpie_shars.py

```diff
@@ -0,0 +1,245 @@
+import torch
+import argparse
+from pathlib import Path
+import torch
+import lightning.pytorch as pl
+from lightning.pytorch.strategies import DDPStrategy
+from torch.utils.data import DataLoader
+from nemo.collections.tts.models import AudioCodecModel
+from lightning.pytorch.callbacks import BasePredictionWriter
+from lhotse import CutSet, MonoCut
+from lhotse.shar.writers.shar import SharWriter
+from lhotse.dataset import SimpleCutSampler
+import time
+
+class SharPredictionWriter(BasePredictionWriter):
+    def __init__(
+        self,
+        output_dir: str,
+        codec_model_name: str,
+        fields: dict,
+        shard_size: int = 1000,
+        buffer_size: int = 128,
+    ) -> None:
+        super().__init__(write_interval="batch")
+
+        self.output_dir = Path(output_dir)
+        self.codec_model_name = codec_model_name
+        self.fields = fields
+        self.shard_size = shard_size
+        self.is_initialized = False
+
+        self.cuts_buffer = list()
+        self.buffer_size = buffer_size
+
+    def setup(self, trainer, pl_module, stage=None):
+        if not self.is_initialized:
+            if trainer.global_rank == 0:
+                self.output_dir.mkdir(parents=True, exist_ok=True)
+                self.shar_writer = SharWriter(
+                    output_dir=str(self.output_dir),
+                    fields=self.fields,
+                    shard_size=self.shard_size,
+                )
+                self.shar_writer.__enter__()
+
+            self.is_initialized = True
+
+    def process_cut_for_saving(self, cut: MonoCut) -> MonoCut:
+        new_custom = {}
+        for k, v in cut.custom.items():
+            if k in self.fields:
+                new_custom[k] = v
+        
+        cut.custom = new_custom
+        return cut
+
+    def write_on_batch_end(self, trainer, pl_module, prediction, batch_indices, batch, batch_idx, dataloader_idx):
+        import time
+        st = time.time()
+        cuts_local, codes_local, context_codes_local = prediction["cuts"], prediction["codes"], prediction["context_codes"]
+        world_size = trainer.world_size
+        if world_size > 1:
+            gathered = [None] * trainer.world_size
+            torch.distributed.all_gather_object(gathered, (cuts_local, codes_local, context_codes_local))
```

Comment:
as suggested, all gather object would sacrifice the GPU utils.

Reply:
I did not understand this suggestion - Should we do something else here?